src/EditorFeatures/Core.Cocoa/Options/VisualStudioMacOptionsStorage.cs

```diff
@@ -0,0 +1,322 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.CodeAnalysis.EditorFeatures.Cocoa;
+
+internal static class VisualStudioMacOptionStorage
+{
+    internal readonly struct PropertyName
+    {
+        private readonly string _name;
+        private readonly string? _vbName;
+        private readonly string? _altName;
+
+        public PropertyName(string name, string? vbName = null, string? altName = null)
+        {
+            _name = name;
+            _vbName = vbName;
+            _altName = altName;
+        }
+
+        private static string SubstituteLanguage(string keyName, string? language)
+            => keyName.Replace("%LANGUAGE%", language switch
+            {
+                LanguageNames.CSharp => "CSharp",
+                LanguageNames.VisualBasic => "VisualBasic",
+                _ => language // handles F#, TypeScript and Xaml
+            });
+
+        public string GetName(string? language)
+            => (_vbName != null && language == LanguageNames.VisualBasic) ? _vbName : SubstituteLanguage(_name, language);
+
+        public string? TryGetAlternativeName(string? language)
+            => _altName != null ? SubstituteLanguage(_altName, language) : null;
+    }
+
+    /// <summary>
+    /// Maps option config name to a legacy property name used in VS Mac XML settings file.
+    /// This list is frozen and only used to enable migration from existing settings files.
+    /// Do not add mappings for new options. <see cref="OptionDefinition.ConfigName"/> is going to be used as a property name for these options.
+    /// </summary>
+    public static readonly IReadOnlyDictionary<string, PropertyName> LegacyPropertyNames = new Dictionary<string, PropertyName>()
+    {
+        {"BlockStructureOptions_CollapseEmptyMetadataImplementationsWhenFirstOpened", new PropertyName("TextEditor.%LANGUAGE%.Specific.CollapseEmptyMetadataImplementationsWhenFirstOpened")},
```

Comment:
This list likely contains more options than VS Mac actually uses. Feel free to suggest which to remove.

Reply:
Yes, let's trim this down to only the options VSMac exposes to users. It is technically possible for users to modify their VSMac settings files to change additional Roslyn options, but this is not supported, and we shouldn't try to migrate it. Those users can update their custom settings in 17.6.

Currently, VSMac serializes the following Roslyn options:

## For C#, F#, and VB.NET

* AutoFormattingOptions
  * AutoFormattingOnCloseBrace
  * AutoFormattingOnReturn
  * AutoFormattingOnSemicolon
  * AutoFormattingOnTyping
* FormattingOptions
  * FormatOnPaste
* GenerationOptions
  * PlaceSystemNamespaceFirst
  * SeparateImportDirectiveGroups
* QuickInfoOptions
  * ShowRemarks
* FeatureOnOffOptions
  * AddImportsOnPaste2
* CompletionOptions
  * ShowItemsFromUnimportedNamespaces
  * TriggerOnDeletion
  * TriggerOnTypingLetters
* CompletionViewOptions
  * ShowCompletionItemFilters
* SolutionCrawlerOptions
  * BackgroundAnalysisScope
* SymbolSearchOptions
  * SearchNuGetPackages

## Once, globally

* FeatureOnOffOptions
  * AutomaticallyCompleteStatementOnSemicolon